### PR TITLE
Implement update & delete of users and other authentication features

### PR DESCRIPTION
**Pull request by GenFusion122**
Added /user/updateUser - sets a new user password for the user making the request
Added /user/deleteUser - deletes the user making the request
Added /user/allUsers - retrieves all users in the database

Admin role is now needed to create a question.
To implement your own Admin role, go to mongoDB and add the ObjectID of the admin role (found in the roles table) to the User you wish to grant the Admin role to.

Resources that are now protected are:
User-service:
- `deleteUser` requires session verification
- `updateUser` requires session verification

Question-service:
- `create` requires admin verification
- `update` requires admin verification
- `delete` requires admin verification
- `deleteAll` requires admin verification
- `findAll` requires session verification
- `findOne` requires session verification

Matching-service:
- `enqueue` requires session verification
- `dequeue` requires session verification

### CODE REVIEW

File: backend/matching-service/controllers/matching.controller.js
Comment (chia-yh) [inline backend/matching-service/controllers/matching.controller.js]: I'm not too familiar with handling sessions with cookies so I just wanted to ask, is this check necessary if the presence of `req.session` and the verification of said session is done in `matching.routes.js`? I don't believe there is any way to directly access the controller, and the check here doesn't seem to verify if `req.session` is valid either.

Also, if this is necessaary, should a similar change be done for `question.controller.js`?

File: backend/matching-service/routes/matching.routes.js
Comment (chia-yh) [inline backend/matching-service/routes/matching.routes.js:27]: Just checking, so this would route the request to the path through `verifySession` before calling `matching.enqueue` right?

File: backend/user-service/middlewares/authJwt.js
Comment (chia-yh) [inline backend/user-service/middlewares/authJwt.js:5]: How this is handled seems rather similar to the `verifySession` and similar methods added for verification in the changes to `matching` and `question`; would it be better to extract such methods into a separate `/middleware(s)` folder in those services as well?

File: backend/user-service/routes/user.routes.js
Comment (chia-yh) [inline backend/user-service/routes/user.routes.js]: Missing a space for `/updateUser`, `/deleteUser` and `/allUsers`

File: frontend/src/app/_helpers/http.interceptor.ts
Comment (chia-yh) [inline frontend/src/app/_helpers/http.interceptor.ts]: Are the new imports used here, or provided to any child of `HttpInterceptor`? I can't seem to see any changes necessitating these imports/additions to the constructor.

File: frontend/src/app/_services/storage.service.ts
Comment (chia-yh) [inline frontend/src/app/_services/storage.service.ts]: See previous comment

File: frontend/src/app/matching/matching.component.ts
Comment (chia-yh) [inline frontend/src/app/matching/matching.component.ts:27]: Changing the timeout from 2s to 30s is a good change

File: frontend/src/app/matching/matching.component.ts
Comment (chia-yh) [inline frontend/src/app/matching/matching.component.ts:50]: Are usernames unique currently? If so, then this change is fine, else it might be a good idea to continue using `id`, but also store the `username` for user readability

File: frontend/src/app/profile/change-password-form.html
Comment (chia-yh) [inline frontend/src/app/profile/change-password-form.html]: Would appreciate @cjyothika 's input on these changes

File: frontend/src/app/profile/profile.component.ts
Comment (chia-yh) [inline frontend/src/app/profile/profile.component.ts:52]: I would consider extracting this into a separate file, perhaps under `/profile/ChangePasswordComponent` for better separation of concerns, and for easier testing

**Review by chia-yh — CHANGES_REQUESTED** [review]
Would like some clarifications on some areas, else I suggested some changes that would be good, but would love to hear thoughts from the team, thanks!

File: frontend/src/app/matching/matching.component.ts
Comment (GenFusion122) [reply frontend/src/app/matching/matching.component.ts:50]: Usernames are indeed unique based on the logic when creating a new user.

File: backend/matching-service/controllers/matching.controller.js
Comment (GenFusion122) [reply backend/matching-service/controllers/matching.controller.js]: Yup you're right, I thought of this method first to check at the controller but I decided to go with the other method to force all tokens to be verified first rather than just trusting the cookie. This was me forgetting to remove this older implementation, my bad.

File: backend/matching-service/routes/matching.routes.js
Comment (GenFusion122) [reply backend/matching-service/routes/matching.routes.js:27]: Yup verifySession would in fact check with the middleware in the user-service to see if the jwt token is valid.

File: backend/user-service/middlewares/authJwt.js
Comment (GenFusion122) [reply backend/user-service/middlewares/authJwt.js:5]: You mean having a separate folder for middlewares in each service?

File: backend/user-service/routes/user.routes.js
Comment (GenFusion122) [reply backend/user-service/routes/user.routes.js]: I blame 1AM coding, my bad.

File: frontend/src/app/app.module.ts
Comment (chia-yh) [inline frontend/src/app/app.module.ts]: Missed this the first time, but should shift these to a module for material imports and provide in root as in the PR for collab, perhaps in this PR if collab is merged before this one, or in that one if this is merged first

File: frontend/src/app/_helpers/http.interceptor.ts
Comment (GenFusion122) [reply frontend/src/app/_helpers/http.interceptor.ts]: Once again its leftover code, will remove.

File: backend/user-service/middlewares/authJwt.js
Comment (chia-yh) [reply backend/user-service/middlewares/authJwt.js:5]: Yep

File: frontend/src/app/_services/storage.service.ts
Comment (GenFusion122) [reply frontend/src/app/_services/storage.service.ts]: Once again, leftover code. :/

File: frontend/src/app/app.module.ts
Comment (GenFusion122) [reply frontend/src/app/app.module.ts]: Sure, if that's the case then we should probably just merge collab first.

File: frontend/src/app/profile/profile.component.ts
Comment (GenFusion122) [reply frontend/src/app/profile/profile.component.ts:52]: Sure, i'll just create a separate folder for both dialog html's and its CSS since they both share the CSS.

Comment (chia-yh) [reply frontend/src/app/profile/profile.component.ts:52]: Still think separate components are better in the interests of unit testing, but maybe @cjyothika can weigh in on this

Comment (cjyothika) [reply frontend/src/app/profile/profile.component.ts:52]: Honestly, I think it's ok since the CSS is pretty minimal but I would say a better/more standardised way to do this would be using bootstrap modals.

File: frontend/src/app/profile/profile.component.ts
Comment (cjyothika) [inline frontend/src/app/profile/profile.component.ts:91]: Do consider using routerLinks in the button component instead.

**Review by cjyothika — APPROVED** [review]
LGTM. However, UI should probably be standardized but I think we should decide this later.

**Review by chia-yh — APPROVED** [review]

**Pull request merged**
